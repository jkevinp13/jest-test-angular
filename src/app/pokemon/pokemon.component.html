<section class="bc-bg-brand-complementary-03 bc-m-4">
    <h1 class="bc-text-center">Pok√©mon</h1>
<div *ngIf="!pokemon; else elseBlock">
    <h2 class="bc-text-center">Loading...</h2>
</div>

<ng-template #elseBlock>
    <h3 >{{ pokemon?.name | titlecase  }}</h3>
    <img  [src]="pokemon?.sprites?.front_default" [alt]="pokemon?.name">
</ng-template>
</section>