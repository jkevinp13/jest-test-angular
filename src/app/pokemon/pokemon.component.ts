import { Component } from '@angular/core';
import { Pokemon } from '../interfaces/pokemon.interface';
import { PokemonService } from '../services/pokemon.service';

@Component({
  selector: 'app-pokemon',
  templateUrl: './pokemon.component.html',
  styleUrls: ['./pokemon.component.scss'],
})
export class PokemonComponent {
  public pokemon?: Pokemon;

  constructor(private pokemonService: PokemonService) {}

  ngOnInit(): void {
    this.pokemonService.getPokemon(6).subscribe({
      next: (pokemon) => {
        this.pokemon = pokemon;
        // console.log(pokemon);
      },
      error: console.log,
    });
  }
}
