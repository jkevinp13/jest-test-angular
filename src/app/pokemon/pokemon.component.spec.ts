import { ComponentFixture, TestBed } from '@angular/core/testing';
import {
  HttpClientTestingModule,
  HttpTestingController,
} from '@angular/common/http/testing';

import { PokemonComponent } from './pokemon.component';
import { PokemonService } from '../services/pokemon.service';
import { of } from 'rxjs';
import { Pokemon } from '../interfaces/pokemon.interface';

describe('PokemonComponent', () => {
  let component: PokemonComponent;
  let fixture: ComponentFixture<PokemonComponent>;
  let compiled: HTMLElement;
  let service: PokemonService;
  let httpMock: HttpTestingController;

  beforeEach(async () => {
    await TestBed.configureTestingModule({
      declarations: [PokemonComponent],
      imports: [HttpClientTestingModule],
      providers: [PokemonService],
    }).compileComponents();
  });

  beforeEach(() => {
    fixture = TestBed.createComponent(PokemonComponent);
    component = fixture.componentInstance;
    service = TestBed.inject(PokemonService);
    httpMock = TestBed.inject(HttpTestingController);

    fixture.detectChanges();
    compiled = fixture.nativeElement;
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('debe de hacer match con el snapshot', () => {
    expect(compiled.innerHTML).toMatchSnapshot();
  });

  it('debe de mostrar un loading al inicio', () => {
    const h2 = compiled.querySelector('h2');
    expect(h2?.textContent).toContain('Loading...');
  });

  it('debe de cargar a charizard inmediatamente', () => {
    const dummyPokemon = {
      name: 'charizardo!!',
      sprites: {
        front_default: 'https://charizard.com/sprite.png',
      },
    };

    const request = httpMock.expectOne('https://pokeapi.co/api/v2/pokemon/6');
    expect(request.request.method).toBe('GET');
    request.flush(dummyPokemon);

    fixture.detectChanges();
    // console.log(compiled.innerHTML);
    const h3 = compiled.querySelector('h3');
    const img = compiled.querySelector('img');

    expect(h3?.textContent?.toLocaleLowerCase()).toContain(
      dummyPokemon.name.toLocaleLowerCase(),
    );
    expect(img?.src).toBe(dummyPokemon.sprites.front_default);
    expect(img?.alt).toBe(dummyPokemon.name);
  });

  it('debe de cargar a charizard inmediatamente con jestSpyOn', () => {
    const dummyPokemon = {
      name: 'charizardo!!',
      sprites: {
        front_default: 'https://charizard.com/sprite.png',
      },
    };
    const spyService = jest
      .spyOn(service, 'getPokemon')
      .mockReturnValue(of(dummyPokemon as Pokemon));
    component.ngOnInit();

    fixture.detectChanges();
    // console.log(compiled.innerHTML);
    const h3 = compiled.querySelector('h3');
    const img = compiled.querySelector('img');

    expect(spyService).toHaveBeenCalledWith(6);
    expect(h3?.textContent?.toLocaleLowerCase()).toContain(
      dummyPokemon.name.toLocaleLowerCase(),
    );
    expect(img?.src).toBe(dummyPokemon.sprites.front_default);
    expect(img?.alt).toBe(dummyPokemon.name);
  });
});
