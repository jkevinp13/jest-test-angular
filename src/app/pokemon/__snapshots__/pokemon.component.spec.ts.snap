// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokemonComponent debe de hacer match con el snapshot 1`] = `
"<section class="bc-bg-brand-complementary-03 bc-m-4"><h1 class="bc-text-center">Pok√©mon</h1><div><h2 class="bc-text-center">Loading...</h2></div><!--bindings={
  "ng-reflect-ng-if-else": "[object Object]"
}--><!--container--></section>"
`;
